### PR TITLE
Fix debug_print() with empty Variable

### PR DESCRIPTION
**Pull request by hvy**
Fixes an error that is raised when calling `debug_print()` on a `Variable` object without `data` (https://github.com/chainer/chainer/issues/3994).

### CODE REVIEW

File: chainer/variable.py [hunk 1]
```diff
@@ -515,6 +515,15 @@ def debug_print(self):
         with cuda.get_device_from_array(self.data) as dev:
             xp = numpy if int(dev) == -1 else cuda.cupy
 
+            if self.data is None:
+                backend = shape = dtype = stats = None
```
Comment (kmaehashi) [inline chainer/variable.py]: * Does it make sense to print `device: CPU` when data is None?
* Please add a test case for this.

Comment (hvy) [reply chainer/variable.py]: - Changed to `device: None`.
- Added test.

**Comment by hvy** [comment]
@kmaehashi san, thanks for the comments. I made changes accordingly (and simplified the code a bit). Could you take another look?

File: chainer/variable.py [hunk 2]
```diff
@@ -499,38 +499,38 @@ def debug_print(self):
...
 - grad: {grad}"""
 
         stats_msg = 'mean={0:.8f}, std={1:.8f}'
 
-        try:
-            device = self.data.device
-        except AttributeError:
-            device = 'CPU'
-
-        with cuda.get_device_from_array(self.data) as dev:
```
Comment (kmaehashi) [inline chainer/variable.py]: Why is this device context removed?
I think this is needed when the `self.data` is on GPU other than the current GPU.

Comment (hvy) [reply chainer/variable.py]: Thanks for pointing that out. I was not aware of this requirement. Fixed.

**Comment by hvy** [comment]
Ready for another review.

**Comment by hvy** [comment]
Ready for another review. Fyi, I squad all commits.

File: chainer/variable.py [hunk 3]
```diff
@@ -499,38 +499,42 @@ def debug_print(self):
...
+                stats = stats_msg.format(float(xp.mean(data)),
+                                         float(xp.std(data)))
+            shape = getattr(data, 'shape', None)
+            dtype = getattr(data, 'dtype', None)
+
             if self.grad is None:
                 grad = None
             elif xp.all(self.grad == 0):
-                grad = 0
+                grad = 0.0
```
Comment (toslunar) [inline chainer/variable.py]: Is this change related to the bugfix?

Comment (hvy) [reply chainer/variable.py]: The motivation was to unify the type with mean and std, but it does not have to be included in this PR as you point out. If this grad needs to be an int, I can change it back.

Comment (hvy) [reply chainer/variable.py]: It probably shouldn't be included in this PR anyhow. Reverted the change. Still, please let me know if there is a particular reason to keep it this way.

**Comment by hvy** [comment]
Rebased.

**Comment by kmaehashi** [comment]
LGTM!
